src/ConnectedMode.UnitTests/ServerEventFilteredChannelTests.cs

```diff
@@ -0,0 +1,152 @@
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.ConnectedMode.ServerSentEvents.Issues;
+using SonarLint.VisualStudio.ConnectedMode.ServerSentEvents.TaintVulnerabilities; 
+using SonarLint.VisualStudio.Integration.UnitTests;
+using SonarQube.Client.Models.ServerSentEvents.ClientContract;
+
+namespace SonarLint.VisualStudio.ConnectedMode.UnitTests
+{
+    // since TaintServerEventChannel and IssueChangedServerEventChannel only differ in the value of the type parameter and in mef export contract, the tests results are the same for both
+    [TestClass]
+    public class ServerEventFilteredChannelTests
+    {
+        [TestMethod]
+        public void TaintAndIssuesMefCtor_CheckIsExported()
+        {
+            MefTestHelpers.CheckTypeCanBeImported<TaintServerEventChannel, ITaintServerEventSource>();
+            MefTestHelpers.CheckTypeCanBeImported<TaintServerEventChannel, ITaintServerEventSourcePublisher>();
+            MefTestHelpers.CheckTypeCanBeImported<IssueChangedServerEventChannel, IIssueChangedServerEventSource>();
+            MefTestHelpers.CheckTypeCanBeImported<IssueChangedServerEventChannel, IIssueChangedServerEventSourcePublisher>();
+        }
+
+        [TestMethod]
+        public async Task Get_AfterPublish_ReturnsCorrectValue()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            testSubject.Publish(serverEvent);
+            var receivedEvent = await testSubject.GetNextEventOrNullAsync();
+
+            receivedEvent.Should().BeSameAs(serverEvent);
+        }
+
+        [TestMethod]
+        public async Task Get_AwaitsForPublishBeforeReturning()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            var getTask =  testSubject.GetNextEventOrNullAsync();
+            testSubject.Publish(serverEvent);
+            var receivedEvent = await getTask;
+
+            receivedEvent.Should().BeSameAs(serverEvent);
+        }
+
+        [TestMethod]
+        public async Task Get_ReturnsMultiplePublishedItemsInCorrectOrder()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvents = Enumerable.Range(0, 5).Select(_ => Mock.Of<ITaintServerEvent>()).ToList();
+            var receivedEvents = new List<ITaintServerEvent>();
+
+            foreach (var serverEvent in serverEvents)
+            {
+                testSubject.Publish(serverEvent);
+            }
+            for (var i = 0; i < serverEvents.Count; i++)
+            {
+                receivedEvents.Add(await testSubject.GetNextEventOrNullAsync());
+            }
+
+            receivedEvents.Should().BeEquivalentTo(serverEvents);
+        }
+
+        [TestMethod]
+        public async Task Get_AlreadyDisposedAndNoMoreItems_ReturnsNull()
```

Comment:
Missing a test of what happens when you call Get immediately after Dispose, without calling publish.
```
var testSubject = CreateTestSubject();
testSubject.Dispose();
var receivedEvent = await testSubject.GetNextEventOrNullAsync();
```

Reply:
I added it, just forgot to reply to the comment